Support for non-Latin TEMP directory path on Windows

rfomin opened the pull request:
Taken from Chocolate Doom: https://github.com/chocolate-doom/chocolate-doom/blob/4bfbd986b2831945250930c2c7e80896254cab16/src/m_misc.c#L564

rfomin committed:
Support for non-Latin TEMP directory path on Windows

rfomin committed:
return current directory if TEMP is not defined

kraflab commented:
What's the problem exactly? The chocolate doom implementation has both reduced coverage and the shocking possibility of returning `.`

kraflab commented:
What about `GetTempPathW`?

rfomin committed:
use GetTempPathW function

FtZPetruska commented on prboom2/src/SDL/i_system.c:
```diff
@@ -254,11 +254,27 @@ const char *I_DoomExeDir(void)
+    if (result == 0 || result > MAX_PATH)
+    {
+      lprintf(LO_ERROR, "GetTempPath(2) failed\n");
+      tmp_path = strdup(".");
```
If the current directory is used as a temporary directory, unzipping a file in the same directory may fail, e.g.:

- `dsda-doom -iwad doom2 -file ./foo.zip`
- `GetTempPathW` fails and `I_GetTempDir` returns `.`
- `D_AddZip` uses `./foo.zip/` as the output directory:
```c
dsda_InitString(&temporary_directory, I_GetTempDir());
dsda_StringCatF(&temporary_directory, "/%s/", dsda_BaseName(zipped_file_name));
M_MakeDir(temporary_directory.string, true);
```

Maybe strip the `.zip` part of the name with `dsda_CutExtension`.

kraflab replied:
We can't use `.` ever, this would be completely insane if files just get dumped into a directory without being requested by a user.

kraflab replied:
Could you leave this branch as an `I_Error` for this PR? I will set up something in the dsda_doom_data directory as a fallback, but raising an error for now is fine with me.

rfomin replied:
Done in https://github.com/kraflab/dsda-doom/pull/304/commits/f4c81bc197c81bc7072445122ef5ce8dd27d7780

kraflab commented on prboom2/src/SDL/i_system.c:
```diff
@@ -254,11 +254,27 @@ const char *I_DoomExeDir(void)
 const char* I_GetTempDir(void)
 {
-  static char tmp_path[PATH_MAX] = {0};
+  static const char* tmp_path = NULL;
```
`static` values are always zero-initialized, so you can leave off the assignment.

rfomin committed:
remove GetTempPath2W

GetTempPath2W is only available starting with Windows 10, I think it's not worth
it check this dynamically with GetProcAddress() etc.

rfomin committed:
call I_Error() if GetTempPathW() failed

kraflab commented:
Awesome, thanks 👍
